### PR TITLE
Get started with Boost Unit Test Framework in Search package

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

**Comment by Rombur** [comment]
What's the plan for this PR? Is it just an example or will it collect all the new tests?

**Comment by aprokop** [comment]
@Rombur It will collect all tests in `Search` package. This is just Damien being kind and showing me an example of how to do that.

**Comment by dalg24** [comment]
I would merge as is and remove the hello world test later.

I started migrating the `Boost{Geometry,Range}Adapters` test which will include examples on how to compare floating points with tolerances and compare sequences of elements.  I will open a new PR for these if this one get merged before.

File: packages/Search/test/tstException.cpp [hunk 1]
```diff
@@ -8,77 +8,21 @@
...
-    catch ( ... )
-    {
-        TEST_ASSERT( 0 );
-    }
+    using namespace DataTransferKit;
+    BOOST_CHECK_THROW( DTK_SEARCH_ASSERT( false ), SearchException );
+    std::string const prefix = "DTK Search exception: ";
+    std::string const message = "Keep calm and chive on!";
+    BOOST_CHECK_EXCEPTION( throw SearchException( message ), SearchException,
+                           [&]( SearchException const &e ) {
```
Comment (dalg24) [inline packages/Search/test/tstException.cpp]: On second thought we might want to check that it derives from `std::exception` here.

File: packages/Search/test/CMakeLists.txt [hunk 2]
```diff
@@ -11,11 +11,22 @@ TRIBITS_ADD_EXECUTABLE_AND_TEST(
   )
 TRIBITS_ADD_EXECUTABLE_AND_TEST(
   Exception
-  SOURCES tstException.cpp unit_test_main.cpp
+  SOURCES tstException.cpp
   COMM serial mpi
```
Comment (dalg24) [inline packages/Search/test/CMakeLists.txt]: @aprokop I overlooked that one when I reviewed #543 
I need to add `NUM_MPI_PROCS 1`.

File: packages/Search/test/CMakeLists.txt [hunk 3]
```diff
@@ -11,11 +11,22 @@ TRIBITS_ADD_EXECUTABLE_AND_TEST(
...
-  STANDARD_PASS_OUTPUT
-  FAIL_REGULAR_EXPRESSION "data race;leak;runtime error"
+  ADDED_EXE_TARGET_NAME_OUT dbc_exe
   )
+# TODO link Boost::dynamic_linking interface target to enable dynamic linking
+# (adds BOOST_ALL_DYN_LINK)
+target_compile_definitions(${dbc_exe} PRIVATE BOOST_TEST_MAIN BOOST_TEST_DYN_LINK)
```
Comment (sslattery) [inline packages/Search/test/CMakeLists.txt]: so each of these `target_compile_definitions` links differently and ultimately changes how `main()` is generated. are you trying to decide here which one you want to use?

File: packages/Search/test/tstException.cpp [hunk 4]
```diff
@@ -8,77 +8,21 @@
...
-        {
-            TEST_ASSERT( 0 );
-        }
-    }
-    catch ( ... )
-    {
-        TEST_ASSERT( 0 );
-    }
+    using namespace DataTransferKit;
+    BOOST_CHECK_THROW( DTK_SEARCH_ASSERT( false ), SearchException );
```
Comment (aprokop) [inline packages/Search/test/tstException.cpp]: Check that the thrown message has the right prefix.

File: packages/Search/test/tstException.cpp [hunk 5]
```diff
@@ -8,77 +8,21 @@
...
+    std::string const prefix = "DTK Search exception: ";
+    std::string const message = "Keep calm and chive on!";
+    BOOST_CHECK_EXCEPTION( throw SearchException( message ), SearchException,
+                           [&]( SearchException const &e ) {
+                               return prefix + message == e.what();
+                           } );
```
Comment (aprokop) [inline packages/Search/test/tstException.cpp]: Add `BOOST_CHECK_NO_EXCEPTION(DTK_SEARCH_ASSERT(true));`

File: packages/Search/test/CMakeLists.txt (hunk 2 above)
Comment (dalg24) [reply packages/Search/test/CMakeLists.txt]: Done

File: packages/Search/test/tstException.cpp (hunk 5 above)
Comment (dalg24) [reply packages/Search/test/tstException.cpp]: Done.

File: packages/Search/test/CMakeLists.txt (hunk 3 above)
Comment (dalg24) [reply packages/Search/test/CMakeLists.txt]: > so each of these `target_compile_definitions` links differently and ultimately changes how `main()` is generated.

exactly

> are you trying to decide here which one you want to use?

I will use all versions depending on the unit tests.  I might also link several unit tests into one executable.  Still figuring that one out.

File: packages/Search/test/tstException.cpp (hunk 4 above)
Comment (dalg24) [reply packages/Search/test/tstException.cpp]: Done.

File: packages/Search/test/CMakeLists.txt [hunk 6]
```diff
@@ -11,11 +11,23 @@ TRIBITS_ADD_EXECUTABLE_AND_TEST(
...
+  ADDED_EXE_TARGET_NAME_OUT dbc_exe
+  )
+# TODO link Boost::dynamic_linking interface target to enable dynamic linking
+# (adds BOOST_ALL_DYN_LINK)
+target_compile_definitions(${dbc_exe} PRIVATE BOOST_TEST_MAIN BOOST_TEST_DYN_LINK)
+
+TRIBITS_ADD_EXECUTABLE_AND_TEST(
+  HelloWorld
+  SOURCES tstHelloWorld.cpp utf_main.cpp
+  COMM serial mpi
```
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Search/test/CMakeLists.txt]: Nah this is a parallel hello world :)

File: packages/Search/test/tstException.cpp [hunk 7]
```diff
@@ -8,77 +8,30 @@
...
-    }
-    catch ( ... )
-    {
-        TEST_ASSERT( 0 );
-    }
-
-    const std::string true_message( "DTK Search exception: cond" );
-    TEST_ASSERT( 0 == message.compare( true_message ) );
-}
+#define BOOST_TEST_MODULE DesignByContract
```
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Search/test/tstException.cpp]: Contract<something>?  Maybe enforcement.  Let's not worry about it now and adjust later.

File: packages/Search/test/tstException.cpp [hunk 8]
```diff
@@ -8,77 +8,30 @@
...
-        TEST_ASSERT( 0 );
-    }
+    using namespace DataTransferKit;
+    BOOST_CHECK_NO_THROW( DTK_SEARCH_ASSERT( true ) );
+    std::string const prefix = "DTK Search exception: ";
+    BOOST_CHECK_EXCEPTION(
+        DTK_SEARCH_ASSERT( false ), SearchException,
+        [&]( std::exception const &e ) {
+            std::string const message = e.what();
+            bool const message_starts_with_prefix = message.find( prefix ) == 0;
```
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Search/test/tstException.cpp]: The string "starts with" the prefix and therefore finds it at position zero.

File: packages/Search/test/tstException.cpp [hunk 9]
```diff
@@ -8,77 +8,30 @@
...
-    }
-
-    const std::string true_message( "DTK Search exception: cond" );
-    TEST_ASSERT( 0 == message.compare( true_message ) );
-}
+#define BOOST_TEST_MODULE DesignByContract
 
-// Test the assertion check
-TEUCHOS_UNIT_TEST( SearchException, assertion_test )
+BOOST_AUTO_TEST_CASE( dumb )
```
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Search/test/tstException.cpp]: Meh.  Could adjust later.  What else do you want?

**Comment by aprokop** [comment]
retest this please

**Pull request merged**